applicationset/controllers/applicationset_controller.go

```diff
@@ -1312,4 +1316,67 @@ func syncApplication(application argov1alpha1.Application, prune bool) (argov1al
 	return application, nil
 }
 
+var ownsHandler = predicate.Funcs{
+	CreateFunc: func(e event.CreateEvent) bool {
+		// if we are the owner and there is a create event, we most likely created it and do not need to
+		// re-reconcile
+		log.Debugln("received create event from owning an application")
+		return false
+	},
+	DeleteFunc: func(e event.DeleteEvent) bool {
+		log.Debugln("received delete event from owning an application")
+		return true
+	},
+	UpdateFunc: func(e event.UpdateEvent) bool {
+		log.Debugln("received update event from owning an application")
+		appOld, isApp := e.ObjectOld.(*argov1alpha1.Application)
+		if !isApp {
+			return false
+		}
+		appNew, isApp := e.ObjectNew.(*argov1alpha1.Application)
+		if !isApp {
+			return false
+		}
+		requeue := shouldRequeueApplicationSet(appOld, appNew)
+		log.Debugf("requeue: %t caused by application %s\n", requeue, appNew.Name)
+		return requeue
+	},
+	GenericFunc: func(e event.GenericEvent) bool {
+		log.Debugln("received generic event from owning an application")
+		return true
+	},
+}
+
+// shouldRequeueApplicationSet determines when we want to requeue an ApplicationSet for reconciling based on an owned
+// application change
+// The applicationset controller owns a subset of the Application CR.
+// We do not need to re-reconcile if parts of the application change outside the applicationset's control.
+// An example being, Application.ApplicationStatus.ReconciledAt which gets updated by the application controller.
+// Additionally, Application.ObjectMeta.ResourceVersion and Application.ObjectMeta.Generation which are set by K8s.
+func shouldRequeueApplicationSet(appOld *argov1alpha1.Application, appNew *argov1alpha1.Application) bool {
+	if appOld == nil || appNew == nil {
+		return false
+	}
+
+	// the applicationset controller owns the application spec, labels, annotations, and finalizers on the applications
+	if !reflect.DeepEqual(appOld.Spec, appNew.Spec) ||
+		!reflect.DeepEqual(appOld.ObjectMeta.GetAnnotations(), appNew.ObjectMeta.GetAnnotations()) ||
+		!reflect.DeepEqual(appOld.ObjectMeta.GetLabels(), appNew.ObjectMeta.GetLabels()) ||
+		!reflect.DeepEqual(appOld.ObjectMeta.GetFinalizers(), appNew.ObjectMeta.GetFinalizers()) {
+		return true
+	}
+
+	// progressive syncs use the application status for updates. if they differ, requeue to trigger the next progression
+	if appOld.Status.Health.Status != appNew.Status.Health.Status || appOld.Status.Sync.Status != appNew.Status.Sync.Status {
+		return true
+	}
+
+	if appOld.Status.OperationState != nil && appNew.Status.OperationState != nil &&
+		appOld.Status.OperationState.Phase != appNew.Status.OperationState.Phase {
```

Comment:
I think `Status.OperationState.StartedAt` also needs to be examined here.

Reply:
Does something outside the appset controller touch that value? If not, I don't think it makes sense to requeue an application since the appset would've just updated it.

Reply:
`Status.OperationState.StartedAt` is handled by the application-controller exclusively, it's not modified by the applicationset-controller. I'm reading it inside the applicationset-controller in order to identify when an Application sync has occurred.